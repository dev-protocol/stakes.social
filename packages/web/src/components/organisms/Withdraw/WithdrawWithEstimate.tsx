import React, { useEffect, useState, useMemo } from 'react'
import {
  useGetMyStakingAmount,
  useGetMyStakingRewardAmount,
  useGetEstimateGas4WithdrawStakingAmount
} from 'src/fixtures/dev-kit/hooks'
import { toBigNumber, whenDefined, whenDefinedAll } from 'src/fixtures/utility'
import { useGetEthPrice } from 'src/fixtures/uniswap/hooks'
import { FormContainer } from 'src/components/molecules/TransactForm/FormContainer'
import { Estimated } from 'src/components/molecules/TransactForm/Estimated'
import { EstimatedGasFeeCard } from 'src/components/molecules/EstimatedGasFeeCard'
import { Withdraw } from '.'

interface Props {
  className?: string
  title?: string
  propertyAddress: string
}

export const WithdrawWithEstimate = ({ className, title, propertyAddress }: Props) => {
  const [withdrawAmount, setWithdrawAmount] = useState<string>('')
  const [claimedTokens, setClaimedTokens] = useState<string | undefined>()
  const [interestTokens, setInterestTokens] = useState<string | undefined>()
  const [withdrawableTokens, setWithdrawableTokens] = useState<string>('')
  const { dev: myStakingRewardAmount } = useGetMyStakingRewardAmount(propertyAddress)
  const { myStakingAmount } = useGetMyStakingAmount(propertyAddress)
  const { estimateGas } = useGetEstimateGas4WithdrawStakingAmount(propertyAddress, withdrawAmount || '0')
  const { data: ethPrice } = useGetEthPrice()
  const estimateGasUSD = useMemo(
    () => whenDefinedAll([estimateGas, ethPrice], ([gas, eth]) => gas.multipliedBy(eth)),
    [estimateGas, ethPrice]
  )

  const estimatedValue = useMemo(
    () => (
      <p>
        {withdrawableTokens || 0} DEV {claimedTokens && interestTokens ? `(${claimedTokens} + ${interestTokens})` : ''}
      </p>
    ),
    [withdrawableTokens, claimedTokens, interestTokens]
  )

  const onChange = (value: string) => {
    setWithdrawAmount(value)
  }

  useEffect(() => {
    const reward = myStakingRewardAmount
    const staking = whenDefined(myStakingAmount, x =>
      x.isGreaterThan(withdrawAmount || 0) ? toBigNumber(withdrawAmount) : x
    )
    setClaimedTokens(whenDefined(staking, x => x.dp(2).toFixed()))
    setInterestTokens(whenDefined(reward, x => x.dp(2).toFixed()))
    setWithdrawableTokens(staking && reward ? staking.plus(reward).dp(2).toFixed() : '0')
  }, [myStakingRewardAmount, myStakingAmount, withdrawAmount, setClaimedTokens])

  return (
    <FormContainer className={className}>
      <Withdraw propertyAddress={propertyAddress} title={title} onChange={onChange} />
      <Estimated title="Staked and Reward Amount">{estimatedValue}</Estimated>
      <EstimatedGasFeeCard
        estimatedGasFee={estimateGas ? estimateGas.toFixed(6) : '-'}
        estimatedGasFeeUSD={estimateGasUSD ? estimateGasUSD.toFixed(2) : ''}
      />
    </FormContainer>
  )
}
