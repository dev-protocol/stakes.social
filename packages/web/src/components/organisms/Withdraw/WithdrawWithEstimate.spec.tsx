import React from 'react'
import Web3 from 'web3'
import { render } from '@testing-library/react'
import { WithdrawWithEstimate } from './WithdrawWithEstimate'
import 'src/__mocks__/window/matchMedia.mock'
import WalletContext from 'src/context/walletContext'

jest.mock('src/fixtures/dev-kit/hooks')
jest.mock('src/fixtures/uniswap/hooks')
jest.mock('src/fixtures/wallet/hooks')
jest.mock('web3')
jest.mock('@dev/graphql')

describe(`${WithdrawWithEstimate.name}`, () => {
  test('Snapshot', () => {
    const component = render(
      <WalletContext.Provider value={{ web3: new Web3(), setProviders: () => {} }}>
        <WithdrawWithEstimate propertyAddress="propertyAddress" />
      </WalletContext.Provider>
    )
    const tree = component.baseElement
    expect(tree).toMatchSnapshot()
  })
})
