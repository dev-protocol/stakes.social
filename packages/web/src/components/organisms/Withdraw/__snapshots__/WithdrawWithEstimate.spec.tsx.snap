// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithdrawWithEstimate Snapshot 1`] = `
<body>
  <div>
    <div
      class="sc-89297o-0 etzBlt"
    >
      <div
        class="sc-89297o-0 etzBlt"
      >
        <div
          class="sc-1d8aayd-1 bcBbeW"
          style="opacity: 1;"
        >
          <span
            class="ant-input-group-wrapper ant-input-group-wrapper-lg ant-input-search ant-input-search-large ant-input-search-with-button sc-1d8aayd-0 fanYmx"
          >
            <span
              class="ant-input-wrapper ant-input-group"
            >
              <span
                class="ant-input-affix-wrapper ant-input-affix-wrapper-lg"
              >
                <input
                  class="ant-input ant-input-lg"
                  id="withdraw"
                  type="number"
                  value=""
                />
                <span
                  class="ant-input-suffix"
                >
                  <button
                    class="sc-1ttu0x8-0 kmiVBa"
                  >
                    DEV
                  </button>
                </span>
              </span>
              <span
                class="ant-input-group-addon"
              >
                <button
                  class="ant-btn ant-btn-primary ant-btn-lg ant-input-search-button"
                  type="button"
                >
                  <span>
                    Withdraw
                  </span>
                </button>
              </span>
            </span>
          </span>
        </div>
        <div
          class="sc-1al8l63-1 bplZGk"
        >
          <div
            class="sc-1al8l63-0 gZPJKN"
          >
            <span
              aria-label="info-circle"
              class="anticon anticon-info-circle"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="info-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                />
                <path
                  d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
          </div>
          <span
            style="font-size: 0.8em;"
          >
            You will receive all accumulated rewards when withdrawing any amount of staked DEV.
          </span>
        </div>
      </div>
      <div
        class="ant-card ant-card-bordered sc-1lowp9k-0 kiBaFy"
      >
        <div
          class="ant-card-head"
        >
          <div
            class="ant-card-head-wrapper"
          >
            <div
              class="ant-card-head-title"
            >
              Staked and Reward Amount
            </div>
          </div>
        </div>
        <div
          class="ant-card-body"
        >
          <p>
            3000
             DEV 
            (0 + 3000)
          </p>
        </div>
      </div>
      <div
        class="ant-card ant-card-bordered ant-card-small sc-18v7ikh-0 dfjrvZ"
      >
        <div
          class="ant-card-head"
        >
          <div
            class="ant-card-head-wrapper"
          >
            <div
              class="ant-card-head-title"
            >
              Gas Fee (predicted)
            </div>
          </div>
        </div>
        <div
          class="ant-card-body"
        >
          <p>
            0.002233
             ETH
            <span
              class="sc-18v7ikh-1 kGdgpU"
            >
               $0.52
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
